File: typescript/processor/src/dsti/intent-density-calculator.ts
/**
 * @docspec:module id="docspec-ts-isd-calculator" name="DSTI Intent Density Calculator"
 * @docspec:description "Calculates the Intent Signal Density (ISD) score for a method
 *   based on all 13 DSTI channels. Score ranges from 0.0 (no intent signals) to 1.0
 *   (maximum density). Uses weighted sum with standardized channel weights that are
 *   consistent across all 6 language implementations."
 * @docspec:boundary "processor-internal"
 * @docspec:since "3.0.0"
 * @docspec:deterministic true
 */
import type { ExtractedSignals } from "./intent-extractor.js";

/**
 * Calculates the Intent Signal Density (ISD) score for a method based on
 * the collected intent signals across all 13 channels. The score ranges
 * from 0.0 (no intent signals) to 1.0 (maximum intent density).
 *
 * Port of the Java IntentDensityCalculator to TypeScript.
 *
 * ISD formula (weights sum to 1.0):
 *   ISD = w1*nameSemantics + w2*guardClauses + w3*branches + w4*dataFlow
 *       + w5*returnType + w6*loops + w7*errorHandling + w8*constants
 *       + w9*nullChecks + w10*assertions + w11*logging + w12*dependencies
 *       + w13*validationAnnotations
 */

// Channel weights (sum = 1.0)
const W_NAME_SEMANTICS = 0.15;
const W_GUARD_CLAUSES = 0.10;
const W_BRANCHES = 0.10;
const W_DATA_FLOW = 0.05;
const W_RETURN_TYPE = 0.05;
const W_LOOPS = 0.08;
const W_ERROR_HANDLING = 0.10;
const W_CONSTANTS = 0.05;
const W_NULL_CHECKS = 0.07;
const W_ASSERTIONS = 0.07;
const W_LOGGING = 0.05;
const W_DEPENDENCIES = 0.08;
const W_VALIDATION_ANNOTATIONS = 0.05;

export class IntentDensityCalculator {
  /**
   * Calculates the ISD score for the given intent signals.
   *
   * @param signals the collected intent signals for a method
   * @returns a score between 0.0 and 1.0
   */
  calculate(signals: ExtractedSignals): number {
    let score = 0.0;

    // Channel 1: Name semantics (0 or W_NAME_SEMANTICS)
    if (signals.nameSemantics?.intent && signals.nameSemantics.intent !== "unknown") {
      score += W_NAME_SEMANTICS;
    }

    // Channel 2: Guard clauses (scaled up to W_GUARD_CLAUSES)
    if (signals.guardClauses && signals.guardClauses > 0) {
      score += Math.min(W_GUARD_CLAUSES, signals.guardClauses * 0.035);
    }

    // Channel 3: Branches (scaled up to W_BRANCHES)
    if (signals.branches && signals.branches > 0) {
      score += Math.min(W_BRANCHES, signals.branches * 0.025);
    }

    // Channel 4: Data flow (0 or W_DATA_FLOW)
    if (signals.dataFlow) {
      const hasReads = signals.dataFlow.reads && signals.dataFlow.reads.length > 0;
      const hasWrites = signals.dataFlow.writes && signals.dataFlow.writes.length > 0;
      if (hasReads && hasWrites) {
        score += W_DATA_FLOW;
      } else if (hasReads || hasWrites) {
        score += W_DATA_FLOW * 0.5;
      }
    }

    // Channel 5: Return type (inferred from verb intent)
    // Query/creation/transformation verbs imply non-void return, adding signal density
    if (signals.nameSemantics?.intent) {
      const intent = signals.nameSemantics.intent;
      if (intent === "query" || intent === "creation" || intent === "transformation") {
        score += W_RETURN_TYPE;
      }
    }

    // Channel 6: Loop properties (up to W_LOOPS)
    if (signals.loopProperties) {
      let loopScore = 0.0;
      if (signals.loopProperties.hasStreams) loopScore += W_LOOPS * 0.6;
      if (signals.loopProperties.hasEnhancedFor) loopScore += W_LOOPS * 0.4;
      score += Math.min(W_LOOPS, loopScore);
    }

    // Channel 7: Error handling (scaled up to W_ERROR_HANDLING)
    if (signals.errorHandling && signals.errorHandling.catchBlocks > 0) {
      score += Math.min(W_ERROR_HANDLING, signals.errorHandling.catchBlocks * 0.035);
    }

    // Channel 8: Constants (scaled up to W_CONSTANTS)
    if (signals.constants && signals.constants.length > 0) {
      score += Math.min(W_CONSTANTS, signals.constants.length * 0.015);
    }

    // Channel 9: Null checks (scaled up to W_NULL_CHECKS)
    if (signals.nullChecks && signals.nullChecks > 0) {
      score += Math.min(W_NULL_CHECKS, signals.nullChecks * 0.02);
    }

    // Channel 10: Assertions (scaled up to W_ASSERTIONS)
    if (signals.assertions && signals.assertions > 0) {
      score += Math.min(W_ASSERTIONS, signals.assertions * 0.025);
    }

    // Channel 11: Logging (scaled up to W_LOGGING)
    if (signals.logStatements && signals.logStatements > 0) {
      score += Math.min(W_LOGGING, signals.logStatements * 0.02);
    }

    // Channel 12: Dependencies (scaled up to W_DEPENDENCIES)
    if (signals.dependencies && signals.dependencies.length > 0) {
      score += Math.min(W_DEPENDENCIES, signals.dependencies.length * 0.025);
    }

    // Channel 13: Validation annotations
    if (signals.validationAnnotations && signals.validationAnnotations > 0) {
      score += Math.min(W_VALIDATION_ANNOTATIONS, signals.validationAnnotations * 0.02);
    }

    return Math.min(1.0, score);
  }

  /**
   * Returns the weight configuration for external inspection / testing.
   */
  getWeights(): Record<string, number> {
    return {
      nameSemantics: W_NAME_SEMANTICS,
      guardClauses: W_GUARD_CLAUSES,
      branches: W_BRANCHES,
      dataFlow: W_DATA_FLOW,
      returnType: W_RETURN_TYPE,
      loops: W_LOOPS,
      errorHandling: W_ERROR_HANDLING,
      constants: W_CONSTANTS,
      nullChecks: W_NULL_CHECKS,
      assertions: W_ASSERTIONS,
      logging: W_LOGGING,
      dependencies: W_DEPENDENCIES,
      validationAnnotations: W_VALIDATION_ANNOTATIONS,
    };
  }
}

File: csharp/processor/Dsti/IntentDensityCalculator.cs
using System;
using System.Collections.Generic;

namespace DocSpec.Analyzer.Dsti;

// [DocModule(Id = "docspec-csharp-isd-calculator", Name = "DSTI Intent Density Calculator")]
// [DocDescription("Calculates the ISD score (0.0-1.0) using 13-channel weighted formula.")]
// [DocSince("3.0.0")]
// [DocDeterministic]

/// <summary>
/// Calculates the Intent Signal Density (ISD) score for a method based on
/// the collected intent signals across all 13 channels. The score ranges
/// from 0.0 (no intent signals) to 1.0 (maximum intent density).
/// The C# equivalent of Java's <c>IntentDensityCalculator</c>.
///
/// <para>ISD formula (weights sum to 1.0):</para>
/// <code>
/// ISD = w1*nameSemantics + w2*guardClauses + w3*branches + w4*dataFlow
///     + w5*returnType + w6*loops + w7*errorHandling + w8*constants
///     + w9*nullChecks + w10*assertions + w11*logging + w12*dependencies
///     + w13*validationAnnotations
/// </code>
/// </summary>
// [DocInvariant(Rules = new[] { "score >= 0.0", "score <= 1.0", "channel weights sum to 1.0" })]
// [DocIntentional("Calculate a weighted intent density score from 13 DSTI channels")]
public class IntentDensityCalculator
{
    // Channel weights (sum = 1.0)
    private const double W_NameSemantics = 0.15;
    private const double W_GuardClauses = 0.10;
    private const double W_Branches = 0.10;
    private const double W_DataFlow = 0.05;
    private const double W_ReturnType = 0.05;
    private const double W_Loops = 0.08;
    private const double W_ErrorHandling = 0.10;
    private const double W_Constants = 0.05;
    private const double W_NullChecks = 0.07;
    private const double W_Assertions = 0.07;
    private const double W_Logging = 0.05;
    private const double W_Dependencies = 0.08;
    private const double W_ValidationAnnotations = 0.05;

    /// <summary>
    /// Calculates the ISD score for the given intent signals.
    /// </summary>
    /// <param name="signals">The collected intent signals for a method.</param>
    /// <returns>A score between 0.0 and 1.0.</returns>
    // [DocMethod(Since = "3.0.0")]
    // [DocDeterministic]
    // [DocPreserves("channel weight proportions")]
    public double Calculate(IntentSignals signals)
    {
        double score = 0.0;

        // Channel 1: Name semantics (0 or W_NameSemantics)
        if (signals.NameSemantics is not null &&
            signals.NameSemantics.Intent is not null &&
            signals.NameSemantics.Intent != "unknown")
        {
            score += W_NameSemantics;
        }

        // Channel 2: Guard clauses (scaled up to W_GuardClauses)
        if (signals.GuardClauses > 0)
        {
            score += Math.Min(W_GuardClauses, signals.GuardClauses * 0.035);
        }

        // Channel 3: Branches (scaled up to W_Branches)
        if (signals.Branches > 0)
        {
            score += Math.Min(W_Branches, signals.Branches * 0.025);
        }

        // Channel 4: Data flow (0, half, or full W_DataFlow)
        if (signals.DataFlow is not null)
        {
            bool hasReads = signals.DataFlow.Reads?.Count > 0;
            bool hasWrites = signals.DataFlow.Writes?.Count > 0;
            if (hasReads && hasWrites)
                score += W_DataFlow;
            else if (hasReads || hasWrites)
                score += W_DataFlow * 0.5;
        }

        // Channel 5: Return type inference from verb
        if (signals.NameSemantics?.Intent is "query" or "creation" or "transformation")
        {
            score += W_ReturnType;
        }

        // Channel 6: Loop properties (up to W_Loops)
        if (signals.LoopProperties is not null)
        {
            double loopScore = 0.0;
            if (signals.LoopProperties.HasLinq) loopScore += W_Loops * 0.6;
            if (signals.LoopProperties.HasForeach) loopScore += W_Loops * 0.4;
            score += Math.Min(W_Loops, loopScore);
        }

        // Channel 7: Error handling (scaled up to W_ErrorHandling)
        if (signals.ErrorHandling is not null && signals.ErrorHandling.CatchBlocks > 0)
        {
            score += Math.Min(W_ErrorHandling, signals.ErrorHandling.CatchBlocks * 0.035);
        }

        // Channel 8: Constants (scaled up to W_Constants)
        if (signals.Constants?.Count > 0)
        {
            score += Math.Min(W_Constants, signals.Constants.Count * 0.015);
        }

        // Channel 9: Null checks (scaled up to W_NullChecks)
        if (signals.NullChecks > 0)
        {
            score += Math.Min(W_NullChecks, signals.NullChecks * 0.02);
        }

        // Channel 10: Assertions (scaled up to W_Assertions)
        if (signals.Assertions > 0)
        {
            score += Math.Min(W_Assertions, signals.Assertions * 0.025);
        }

        // Channel 11: Logging (scaled up to W_Logging)
        if (signals.LogStatements > 0)
        {
            score += Math.Min(W_Logging, signals.LogStatements * 0.02);
        }

        // Channel 12: Dependencies (scaled up to W_Dependencies)
        if (signals.Dependencies?.Count > 0)
        {
            score += Math.Min(W_Dependencies, signals.Dependencies.Count * 0.025);
        }

        // Channel 13: Validation annotations (scaled up to W_ValidationAnnotations)
        if (signals.ValidationAnnotations > 0)
        {
            score += Math.Min(W_ValidationAnnotations, signals.ValidationAnnotations * 0.02);
        }

        return Math.Min(1.0, score);
    }
}

/// <summary>
/// Container for all 13 intent signal channels collected from a method.
/// </summary>
public class IntentSignals
{
    public NamingAnalyzer.NameSemantics? NameSemantics { get; set; }
    public int GuardClauses { get; set; }
    public int Branches { get; set; }
    public DataFlowSignals? DataFlow { get; set; }
    public LoopSignals? LoopProperties { get; set; }
    public ErrorHandlingSignals? ErrorHandling { get; set; }
    public List<string>? Constants { get; set; }
    public int NullChecks { get; set; }
    public int Assertions { get; set; }
    public int LogStatements { get; set; }
    public List<string>? Dependencies { get; set; }
    public int ValidationAnnotations { get; set; }
}

public class DataFlowSignals
{
    public List<string>? Reads { get; set; }
    public List<string>? Writes { get; set; }
}

public class LoopSignals
{
    /// <summary>Whether the method uses LINQ (equivalent to Java Streams).</summary>
    public bool HasLinq { get; set; }
    /// <summary>Whether the method uses foreach loops.</summary>
    public bool HasForeach { get; set; }
    /// <summary>Names of stream/LINQ operations found (e.g., Select, Where, GroupBy).</summary>
    public List<string>? StreamOps { get; set; }
}

public class ErrorHandlingSignals
{
    public int CatchBlocks { get; set; }
    public bool HasFinally { get; set; }
    public List<string>? CaughtTypes { get; set; }
}
